src/main/java/gregtech/api/gui/Widget.java

```diff
@@ -110,13 +115,18 @@ public final Size getSize() {
     }
 
     public boolean isVisible() {
-        return isVisible;
+        return visibilitySupplier != null ? visibilitySupplier.getAsBoolean() : isVisible;
     }
 
     public void setVisible(boolean visible) {
         isVisible = visible;
     }
 
+    public Widget setVisibilitySupplier(BooleanSupplier supplier) {
+        visibilitySupplier = supplier;
+        return this;
+    }
+
```

Comment:
My advice is not to add this API, it is somewhat coupled. I think it's better to check the game state and update the visible in MazeApp's updateScreen, what do you think? Of course, if you think this interface is really necessary and will be used in future widgets, then let it go. I don't think it's a problem.

Reply:
Personally, while I think this could be done, it would require one of two implementations. Either I would loop through the entire array of widgets, and apply the suppliers one-by-one with a variable for each of them, or I would create an instance variable for everything that needed dynamic visibility, and automatically pass in the argument on each cycle. However, in both cases, I would need 11 variables just to implement this for the MazeApp, some of which with very similar names. This would probably be harder to read then just adding a supplier to a widget directly. Also, I will probably be using this in the third game in this project, and it may end up having more uses in the future.